### PR TITLE
feat: Reference editor custom card render prop

### PR DESCRIPTION
**Pull request by DanweDE**
Adds a `renderCard: (props: CustomEntryCardProps) => React.ReactElement | false;` prop to reference editors.

Closes https://github.com/contentful/field-editors/issues/283

### CODE REVIEW

File: packages/reference/src/entries/WrappedEntryCard/FetchingWrappedEntryCard.tsx [hunk 1]
```diff
@@ -13,6 +15,22 @@ export type EntryCardReferenceEditorProps = ReferenceEditorProps & {
...
+export type WrappedEntryCardProps = {
+  getAsset: (assetId: string) => Promise<unknown>;
+  getEntityScheduledActions: SpaceAPI['getEntityScheduledActions'];
+  getEntryUrl?: (entryId: string) => string;
+  isDisabled: boolean;
+  size: 'default' | 'small';
+  localeCode: string;
+  defaultLocaleCode: string;
+  allContentTypes: ContentType[];
+  entry: 'failed' | undefined | Entry; // TODO: Work out correct type for entry
```
Comment (DanweDE) [inline packages/reference/src/entries/WrappedEntryCard/FetchingWrappedEntryCard.tsx:30]: This should never be `'failed'`, making this a public interface would be quite ugly design plus, it seems not even necessary in `WrappedEntryCard`. There is some `entry?.sys.id` usage so maybe there's a valid case for `undefined`, although `FetchingWrappedEntryCard` seems to catch this case for us, so not sure if there's a point at all.

File: packages/reference/src/common/ReferenceEditor.tsx [hunk 2]
```diff
@@ -2,22 +2,21 @@ import * as React from 'react';
...
-   * is the field disabled initially
+   * Whether or not the field should be disabled initially.
    */
   isInitiallyDisabled: boolean;
-
   sdk: FieldExtensionSDK;
-
   viewType: ViewType;
-
+  renderCard?: (props: CustomEntryCardProps) => React.ReactElement | false;
```
Comment (m10l) [inline packages/reference/src/common/ReferenceEditor.tsx:17]: Might make sense to rename `renderCard` to `customEntryCardElement` - feels a bit more descriptive as a prop. What do you think @DanweDE @suevalov ?

File: packages/reference/src/entries/WrappedEntryCard/WrappedEntryCard.tsx [hunk 3]
```diff
@@ -21,18 +21,18 @@ const styles = {
   }),
 };
 
-interface WrappedEntryCardProps {
+export interface WrappedEntryCardProps {
   getEntityScheduledActions: SpaceAPI['getEntityScheduledActions'];
   getAsset: (assetId: string) => Promise<unknown>;
-  getEntryUrl?: (entryId: string) => string;
```
Comment (m10l) [inline packages/reference/src/entries/WrappedEntryCard/WrappedEntryCard.tsx:27]: @DanweDE Would this be a breaking change?

Comment (DanweDE) [reply packages/reference/src/entries/WrappedEntryCard/WrappedEntryCard.tsx:27]: Yes, but only for the `WrappedEntryCard`. One of the commits is marked as breaking and I've updated the RT editor accordingly.

File: packages/reference/src/common/ReferenceEditor.tsx (hunk 2 above)
Comment (DanweDE) [reply packages/reference/src/common/ReferenceEditor.tsx:17]: this is a function returning an element so the card prop should include a verb I think.

Comment (m10l) [inline packages/reference/src/common/ReferenceEditor.tsx:17]:
```suggestion
  renderCustomEntryCard?: (props: CustomEntryCardProps) => React.ReactElement | false;
```

File: packages/reference/src/common/SingleReferenceEditor.tsx [hunk 4]
```diff
@@ -2,14 +2,15 @@ import * as React from 'react';
...
-import { ReferenceEditor, ReferenceEditorProps } from './ReferenceEditor';
+import { CustomEntryCardProps, ReferenceEditor, ReferenceEditorProps } from './ReferenceEditor';
 
 type ChildProps = {
   entityId: string;
   entityType: EntityType;
   isDisabled: boolean;
   setValue: (value: ReferenceValue | null | undefined) => void;
   allContentTypes: ContentType[];
+  renderCard?: (props: CustomEntryCardProps) => React.ReactElement | false;
```
Comment (m10l) [inline packages/reference/src/common/SingleReferenceEditor.tsx:13]:
```suggestion
  renderCustomEntryCard?: (props: CustomEntryCardProps) => React.ReactElement | false;
```

File: packages/reference/src/entries/MultipleEntryReferenceEditor.mdx [hunk 5]
```diff
@@ -100,9 +48,46 @@ import changedEntry from '../__fixtures__/changed_entry.json';
...
+
+<Playground>
+  {() => {
+    const isInitiallyDisabled = window.localStorage.getItem('initialDisabled');
+    const instanceParams = window.localStorage.getItem('instanceParams');
+    const [sdk, mitt] = newReferenceEditorFakeSdk();
+    return (
+      <div data-test-id="multiple-references-editor-custom-cards-integration-test">
+        <MultipleEntryReferenceEditor
+          renderCard={(props) => props.entry.fields.exField ? <Card testId='custom-card'>
```
Comment (m10l) [inline packages/reference/src/entries/MultipleEntryReferenceEditor.mdx:68]:
```suggestion
          renderCustomEntryCard={(props) => props.entry.fields.exField ? <Card testId='custom-card'>
```

File: packages/reference/src/entries/SingleEntryReferenceEditor.mdx [hunk 6]
```diff
@@ -89,23 +48,58 @@ import publishedEntry from '../__fixtures__/published_entry.json';
...
 
+<Playground>
+  {() => {
+    const isInitiallyDisabled = window.localStorage.getItem('initialDisabled');
+    const instanceParams = window.localStorage.getItem('instanceParams');
+    const [sdk, mitt] = newReferenceEditorFakeSdk();
+    return (
+      <div>
+        <SingleEntryReferenceEditor
+          renderCard={(props) => props.entry.fields.exField ? <Card testId='custom-card'>
```
Comment (m10l) [inline packages/reference/src/entries/SingleEntryReferenceEditor.mdx:67]:
```suggestion
          renderCustomEntryCard={(props) => props.entry.fields.exField ? <Card testId='custom-card'>
```

File: packages/reference/src/entries/SingleEntryReferenceEditor.tsx [hunk 7]
```diff
@@ -6,13 +6,14 @@ import { FetchingWrappedEntryCard } from './WrappedEntryCard/FetchingWrappedEntr
 export function SingleEntryReferenceEditor(props: ReferenceEditorProps) {
   return (
     <SingleReferenceEditor {...props} entityType="Entry">
-      {({ allContentTypes, isDisabled, entityId, setValue }) => {
+      {({ allContentTypes, isDisabled, entityId, setValue, renderCard }) => {
```
Comment (m10l) [inline packages/reference/src/entries/SingleEntryReferenceEditor.tsx:9]:
```suggestion
      {({ allContentTypes, isDisabled, entityId, setValue, renderCustomEntryCard }) => {
```

File: packages/reference/src/entries/SingleEntryReferenceEditor.tsx [hunk 8]
```diff
@@ -6,13 +6,14 @@ import { FetchingWrappedEntryCard } from './WrappedEntryCard/FetchingWrappedEntr
...
         return (
           <FetchingWrappedEntryCard
             {...props}
             allContentTypes={allContentTypes}
             isDisabled={isDisabled}
             entryId={entityId}
+            renderCard={renderCard}
```
Comment (m10l) [inline packages/reference/src/entries/SingleEntryReferenceEditor.tsx:16]:
```suggestion
            renderCustomEntryCard={renderCustomEntryCard}
```

File: packages/reference/src/entries/WrappedEntryCard/FetchingWrappedEntryCard.tsx [hunk 9]
```diff
@@ -13,6 +15,7 @@ export type EntryCardReferenceEditorProps = ReferenceEditorProps & {
   isDisabled: boolean;
   onRemove: () => void;
   cardDragHandle?: React.ReactElement;
+  renderCard?: (props: CustomEntryCardProps) => React.ReactElement | false;
```
Comment (m10l) [inline packages/reference/src/entries/WrappedEntryCard/FetchingWrappedEntryCard.tsx:18]:
```suggestion
  renderCustomEntryCard?: (props: CustomEntryCardProps) => React.ReactElement | false;
```

File: packages/reference/src/entries/WrappedEntryCard/FetchingWrappedEntryCard.tsx [hunk 10]
```diff
@@ -80,43 +108,33 @@ export function FetchingWrappedEntryCard(props: EntryCardReferenceEditorProps) {
...
+      ),
+      isDisabled: props.isDisabled,
+      size,
+      localeCode: props.sdk.field.locale,
+      defaultLocaleCode: props.sdk.locales.default,
+      cardDragHandle: props.cardDragHandle,
+      onEdit,
+      onRemove,
+    };
+    if (props.renderCard) {
```
Comment (m10l) [inline packages/reference/src/entries/WrappedEntryCard/FetchingWrappedEntryCard.tsx:125]:
```suggestion
    if (props.renderCustomEntryCard) {
```

File: packages/reference/src/entries/WrappedEntryCard/FetchingWrappedEntryCard.tsx [hunk 11]
```diff
@@ -80,43 +108,33 @@ export function FetchingWrappedEntryCard(props: EntryCardReferenceEditorProps) {
...
+      const renderedCustomCard = props.renderCard(sharedCardProps);
```
Comment (m10l) [inline packages/reference/src/entries/WrappedEntryCard/FetchingWrappedEntryCard.tsx:126]:
```suggestion
      const renderedCustomCard = props.renderCustomEntryCard(sharedCardProps);
```

File: packages/reference/src/common/ReferenceEditor.tsx (hunk 2 above)
Comment (DanweDE) [reply packages/reference/src/common/ReferenceEditor.tsx]: Settling for `renderCustomCard`

**Pull request merged**